core/src/main/scala/chisel3/experimental/hierarchy/Instantiate.scala

```diff
@@ -0,0 +1,191 @@
+  // types that contain Data.
+  // A more complete and robust solution would require either refining hashCode and equality on Data
+  // to have the desired behavior, or using a different Map implementation that uses typeclasses for
+  // equality and hashcode (eg. cats-collections-core's HashMap)
+  private class DataBox(private val d: Data) {
+
+    private def convertDataForHashing(data: Data): Any = data match {
+      // Using toString is a bit weird but it works
+      case elt: Element => elt.toString
+      case rec: Record  =>
+        // Purely structural, actual class of Record isn't involved
+        rec.elements.map { case (name, data) => name -> convertDataForHashing(data) }
+      case vec: Vec[_] =>
+        // We could map on elements but that's a lot of duplicated work for a type
+        // Note that empty vecs of the same type will give same hash value, probably should be equal
+        // as well, but Vec.typeEquivalent checks sample_element so they will not be equal
+        ("Vec", vec.size, vec.headOption.map(convertDataForHashing(_)))
+    }
+
+    override def hashCode: Int = convertDataForHashing(d).hashCode
+
+    // If literals, check same types and equal
+    // If types, chck same types
+    // If bound, fall back to normal equality
+    override def equals(that: Any): Boolean = {
+      that match {
+        case that: DataBox =>
+          // def because it's not needed by non-literal but is synthesizable check
+          def sameTypes = DataMirror.checkTypeEquivalence(this.d, that.d)
+          // Lits are synthesizable so must check them first
+          if (this.d.isLit) {
+            that.d.isLit &&
+            (this.d.litValue == that.d.litValue) &&
+            sameTypes
+          } else if (isSynthesizable(this.d)) {
+            this.d.equals(that.d)
+          } else {
+            // We have checked that this.d is not synthesizable but need to check that.d as well
+            sameTypes && !isSynthesizable(that.d)
+          }
+
+        case _ => false
+      }
+    }
+  }
+
+  // Recursively box all Data (by traversing Products and Iterables) in DataBoxes
+  private def boxAllData(a: Any): Any = a match {
+    case d:  Data        => new DataBox(d) // Must check this before Iterable because Vec is Iterable
+    case it: Iterable[_] => it.map(boxAllData(_))
+    case p:  Product     => p.productIterator.map(boxAllData(_)).toSeq
+    case other => other
+  }
+
+  /** This is not part of the public API, do not call directly! */
+  def _impl[K, A <: BaseModule: ru.WeakTypeTag](
+    args: K,
+    f:    K => A
+  )(
+    implicit sourceInfo: SourceInfo,
+    compileOptions:      CompileOptions
+  ): Instance[A] = {
+    val tag = implicitly[ru.WeakTypeTag[A]]
+    // Include type of module in key since different modules could have the same arguments
+    // CompileOptions need to be included but are likely a source of confusion for users
+    val key = (boxAllData(args), tag, compileOptions)
```

Comment:
I wonder if it is possible to let user being able to override the `key` implementation, like they can rewrite the `hashCode` and `equal` API  ?

Reply:
We could, but I'm curious what the use case is? We should definitely solve the "The DataBox Problem", but other than making sure that `Data` behaves correctly here, what need is there for changing the key?

Reply:
I think only users can decide how to cache the Definition. for example:
1. different parameter might have a same RTL but different metadata, or some parameter happen to calculate to a same result, which might lead to different elaboration to a same RTL.
2. On the other hand. Even a same parameter can lead to different RTL when module depends on some global variables, this cannot be avoid by compiler, only user can decide: every time I initiate this module, force it to re-elaborate.

For previous case, user can define the cache key calculation method to explicitly say: when calculation results are same, don't elaborate it. 
For the second case, If user makes some Defination depending on global variables, we have to provide a control to this case: They can configure the key cache with the global variable they need, or just randomize the key to force re-elaborate.

I think this is the reason why I think the cache key should be a user configurable variable.

Reply:
> I think this is the reason why I think the cache key should be a user configurable variable.

Could the user just not use the Instantiate API in that case? I guess the question is the Instantiate API one way to do this for "well formed" module constructors that don't depend on weird global variables to build themselves, and that do want different things for different passed-in parameters?

Reply:
> Could the user just not use the Instantiate API in that case?

I think it's still reasonable to give user this flexibility, but ask user to give the `cachekey` implementation.  I don't think it's a good solution to provide a "well formed" Chisel module guideline, since there are a lot of code styles, and the born of this API will limit them.
I prefer throwing the "cache key" burden to user, since only the users can know what the cache key is, Jack's implementation provides a neat default implementation, but making some advance users being able to override the `cachekey` can be a good thing to the API usage:
if using this API, users are not allowed to depends on global mutable or obey some code guideline.
vs
if using this API, if users depend on global mutable, or do other thing which is possible to permute the `Module` elaborations, please mix-in a `HasElaborationCache` trait to `Module` to override the default cache key implementation.